test/shared/src/main/scala/zio/test/environment/package.scala

```diff
@@ -335,83 +338,156 @@ package object environment extends PlatformSpecific {
+                      scheduled.offer(end, Sleep(end, latch, fiberId)).as(true)
                     else
-                      (false, data)
-                  }
+                      STM.succeedNow(false)
+                  }.commit
           _ <- if (await) warningStart *> latch.await else latch.succeed(())
         } yield ()
 
       /**
-       * Returns a list of the wall clock times at which all queued effects are
-       * scheduled to resume.
+       * Returns a list of the times at which all queued effects are scheduled
+       * to resume.
        */
       lazy val sleeps: UIO[List[Duration]] =
-        clockState.get.map(_.sleeps.map(_._1))
+        scheduled.toList.map(_.map(_.duration)).commit
 
       /**
        * Returns the time zone.
        */
       lazy val timeZone: UIO[ZoneId] =
-        fiberState.get.map(_.timeZone)
-
-      private def run(wakes: List[(Duration, Promise[Nothing, Unit])]): UIO[Unit] =
-        UIO.foreach(wakes.sortBy(_._1))(_._2.succeed(())).unit
+        clockState.get.map(_.timeZone).commit
 
+      /**
+       * Cancels the warning message that is displayed if a test is using time
+       * but is not advancing the `TestClock`.
+       */
       private[TestClock] val warningDone: UIO[Unit] =
         warningState.updateSome[Any, Nothing] {
           case WarningData.Start          => ZIO.succeedNow(WarningData.done)
           case WarningData.Pending(fiber) => fiber.interrupt.as(WarningData.done)
         }
 
+      /**
+       * Polls untl the specified fiber completes execution or is indefinitely
+       * suspended, meaning that the fiber and all the fibers it is blocking on
+       * and all the fibers they are blocking on recursively are suspended.
+       */
+      private def awaitSuspended(fiberId: Fiber.Id): UIO[Unit] =
+        live.provide {
+          suspended(fiberId).repeat {
+            Schedule.doUntilEquals(true) && Schedule.fixed(1.millisecond)
+          }
+        }.unit
+
+      /**
+       * Constructs a `Duration` from an `OffsetDateTime`.
+       */
+      private def fromDateTime(dateTime: OffsetDateTime): Duration =
+        Duration(dateTime.toInstant.toEpochMilli, TimeUnit.MILLISECONDS)
+
+      /**
+       * Run all effects scheduled to occur on or before the specified
+       * duration, in order.
+       */
+      private def runUntil(duration: STM[Nothing, Duration]): UIO[Unit] =
+        duration.flatMap { duration =>
+          scheduled.peekOption.flatMap {
+            case Some(sleep) if sleep.duration <= duration =>
+              clockState.update(_.copy(duration = sleep.duration)) *>
+                scheduled.take.map(_.copy(duration = duration)).asSome
+            case _ =>
+              clockState.update(_.copy(duration = duration)) *> STM.none
+          }
+        }.commit.flatMap {
+          case None => UIO.unit
+          case Some(sleep) =>
+            sleep.promise.succeed(()) *>
```

Comment:
Would it make sense to `ZIO.yieldNow` here? In most cases the `awaitSuspend` will return `false` the first time, as the sleeping Fiber will not have finished. On a single threaded environment that will happen every time.

Reply:
Sure.